### PR TITLE
Set minthreads on webhost start

### PR DESCRIPTION
**Pull request by paulbatum**
Performance profiling has revealed that the function execution latency is often impacted by thread starvation. This can be mitigated by configuring the minimum number of threads that will be injected by the threadpool in an unthrottled manner.

See the following resources for more context:
https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.setminthreads?view=netframework-4.7
https://docs.microsoft.com/en-us/azure/redis-cache/cache-faq#important-details-about-threadpool-growth
https://blogs.msdn.microsoft.com/perfworld/2010/01/13/how-can-i-improve-the-performance-of-asp-net-by-adjusting-the-clr-thread-throttling-properties/

Let me know if you'd like to see some tests included for this particular change. I'm not sure if a unit test would be useful here. I could write a E2E test that calls a test script that queries threadpool.GetMinThreads to confirm its no longer set to 1. Let me know if you think that would be worthwhile.

## Performance Results 
Most of the profiling has been done for HTTP as this is the scenario that feels the impact of higher latencies the most. In general this change improves latency at the 95th and 99th percentiles with no impact to the min or median.

I also deployed this change to an Event Hub triggered function running at high scale (430 million events processed per hour) with no observed impact to throughput or E2E latency.

Some notes:
- The HTTP data is E2E latency measurements in milliseconds, captured using [artillery.io](https://artillery.io/).
- Dedicated refers to an app service plan with a single core VM.
- The dashboard was turned off

### HTTP 50 RPS constant load

This is the scenario I collected the most samples for. These results are a summary derived by taking the median of 10 different runs per scenario. See [this gist ](https://gist.github.com/paulbatum/5d544a43a60f47b6339cdd7c4157eba2) for full 50 RPS results.

| Scenario                |  Min |  Median |  P95 |  P99 |  Max | 
|-------------------------|------|---------|------|------|------| 
| Dedicated without fix   | 154  | 159     | 970  | 2072 | 3689 | 
| Dedicated with fix      | 154  | 156     | 220  | 1594 | 2720 | 
| Consumption without fix | 153  | 156     | 439  | 1786 | 9171 | 
| Consumption with fix    | 154  | 157     | 170  | 335  | 5892 |

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Global.asax.cs [hunk 1]
```diff
@@ -45,6 +47,25 @@ private static void VerifyAndEnableShadowCopy(WebHostSettings webHostSettings)
...
+        {
+            // For information on MinThreads, see:
+            // https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.setminthreads?view=netframework-4.7
+            // https://docs.microsoft.com/en-us/azure/redis-cache/cache-faq#important-details-about-threadpool-growth
+            // https://blogs.msdn.microsoft.com/perfworld/2010/01/13/how-can-i-improve-the-performance-of-asp-net-by-adjusting-the-clr-thread-throttling-properties/
+            //
+            // This behavior can be overridden by using the "ComPlus_ThreadPool_ForceMinWorkerThreads" environment variable (honored by the .NET threadpool).
+
+            // The dynamic plan has some limits that mean that a given instance is using effectively a single core, so we should not use Environment.Processor count in this case.
+            var effectiveCores = isDynamicSku ? 1 : Environment.ProcessorCount;
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Global.asax.cs]: Is this check necessary? I'm pretty sure that when running with the Job Object configuration setup in the dynamic SKU, the number of cores will be correctly reported (1) regardless of the machine size. Kranthi can confirm.

Comment (paulbatum) [inline src/WebJobs.Script.WebHost/Global.asax.cs]: Its necessary. Here's some code I just ran on a consumption function app (without this fix), and its output:

```csharp
public static void Run(HttpRequestMessage req, TraceWriter log)
{
    log.Info(Environment.ProcessorCount.ToString());
    log.Info("C# HTTP trigger function processed a request.");

    System.Threading.ThreadPool.GetMinThreads(out int x, out int y);
    log.Info($"{x},{y}");
}
```

```
2017-07-19T15:53:07.150 Function started (Id=c3797636-6f5a-4ce4-abd7-eb43b38f9c3c)
2017-07-19T15:53:07.150 2
2017-07-19T15:53:07.150 C# HTTP trigger function processed a request.
2017-07-19T15:53:07.150 1,1
2017-07-19T15:53:07.150 Function completed (Success, Id=c3797636-6f5a-4ce4-abd7-eb43b38f9c3c, Duration=3ms)
```

I checked the logs and this code happened to execute on a medium VM.

Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Global.asax.cs]: Sounds good. I had tested with similar code and gotten the expected result, so this is a bit concerning, but not related to this changes.

**Comment by fabiocav** [comment]
@paulbatum do you have any changes remaining? Or is this good to merge?

**Pull request merged**